_posts/2021-02-02-enso-prediction/enso-prediction.Rmd

```diff
@@ -0,0 +1,2410 @@
+      phase_code PHASE count   avg
+           <dbl> <chr> <int> <dbl>
+    1          1 E        93  28.1
+    2          2 L        93  25.9
+    3          3 N       182  27.2
+
+This doesn't look too bad. Of course, we again see the overall rise in temperature, irrespective of phase.
+
+Lastly, we normalize the index, same as we did for the input data.
+
+```{r}
+train_mean_nino <- mean(nino_train$NINO34_MEAN)
+train_sd_nino <- sd(nino_train$NINO34_MEAN)
+
+nino_train <- nino_train %>% mutate(NINO34_MEAN = scale(NINO34_MEAN, center = train_mean_nino, scale = train_sd_nino))
+nino_valid <- nino_valid %>% mutate(NINO34_MEAN = scale(NINO34_MEAN, center = train_mean_nino, scale = train_sd_nino))
+```
+
+On to the `torch` dataset.
+
+# `Torch` dataset
+
+The dataset is responsible for correctly matching up inputs and targets.
+
+Our goal is to take six months of global sea surface temperatures and predict the Niño 3.4 Index for the following month. Input-wise, the model will expect the following format semantics:
+
+`batch_size * timesteps * width * height * channels`, where
+
+-   `batch_size` is the number of observations worked on in one round of computations,
+
+-   `timesteps` chains consecutive observations from adjacent months,
+
+-   `width` and `height` together constitute the spatial grid, and
+
+-   `channels` corresponds to available visual channels in the "image".
+
+In `.getitem()`, we select the consecutive observations, starting at a given index, and stack them in dimension one. (One, not two, as batches will only start to exist once the `dataloader` comes into play.)
+
+Now, what about the target? Our ultimate goal was -- is -- predicting the Niño 3.4 Index. However, as you see we define three targets: One is the index, as expected; an additional one holds the spatially-gridded sea surface temperatures for the prediction month. Why? Our main instrument, the most prominent constituent of the model, will be a convLSTM, an architecture designed for *spatial* prediction. Thus, to train it efficiently, we want to give it the opportunity to predict values on a spatial grid. So far so good; but there's one more target, the phase/category. This was added for experimentation purposes: Maybe predicting *both* index *and* phase helps in training?
+
+Finally, here is the code for the dataset. In our experiments, we based predictions on inputs from the preceding six months (`n_timesteps <- 6`). This is a parameter you might want to play with, though.
+
+```{r}
+n_timesteps <- 6
+
+enso_dataset <- dataset(
+  name = "enso_dataset",
+  
+  initialize = function(sst, nino, n_timesteps) {
+    self$sst <- sst
+    self$nino <- nino
+    self$n_timesteps <- n_timesteps
+  },
+  
+  .getitem = function(i) {
+    x <- torch_tensor(self$sst[, , i:(n_timesteps + i - 1)]) # (360, 115, n_timesteps)
+    x <- torch_split(x, 1, dim = 3) # list of length n_timesteps of tensors (360, 115, 1)
+    x <- torch_stack(x) # (n_timesteps, 360, 115, 1)
```

Comment:
is this, just for transposing? why not `torch_transpose`?

Reply:
how do you mean transpose? I mean, I need to create the timesteps/sequence dimension, as for RNNs, - how could I do that with `torch_transpose()` instead of `torch_stack()`?

Reply:
It seems that you split into one dimension and then stack in another just to swap the axis. I meant something like this:

``` r
library(torch)
x <- torch_randn(360, 115, 16)
s <- torch_split(x, 1, dim = 3)
torch_allclose(
  torch_stack(s),
  x$permute(c(3,1,2))$unsqueeze(4)  
)
#> [1] TRUE
```

Reply:
finally got it :-) thanks!